Add a github action step to run the tests against current master

thufschmitt opened the pull request:
Make github actions also run the testsuite using a `nix-daemon` executable that comes from the master branch.

This allows ensuring that every commit on master is backwards-compatible with the previous one.

thufschmitt committed:
Add a github action step to run the tests against current master

edolstra commented on .github/workflows/test.yml:
```diff
@@ -21,6 +21,9 @@ jobs:
     - run: nix-build -A checks.$(nix-instantiate --eval -E '(builtins.currentSystem)')
+    - run: |
+      NIX_BUILD=$(nix-build -A defaultPackage.$(nix-instantiate --eval -E '(builtins.currentSystem)'))/bin/nix-build
+      $NIX_BUILD --option experimental-features 'flakes' --impure --expr 'builtins.getFlake (builtins.toPath ./.).lib.testAgainst.x86_64-linux (builtins.getFlake (builtins.toPath ./.)).defaultPackage.x86_64-linux'
```
Why is `--impure` needed?

regnat replied:
Both flake inputs (`./.` and `github:nixos/nix`) are mutable refs, so they aren’t allowed in pure mode

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-16/14379/1

tomberek commented on .github/workflows/test.yml:
```diff
@@ -21,6 +21,9 @@ jobs:
     - run: nix-build -A checks.$(nix-instantiate --eval -E '(builtins.currentSystem)')
+    - run: |
+      NIX_BUILD=$(nix-build -A defaultPackage.$(nix-instantiate --eval -E '(builtins.currentSystem)'))/bin/nix-build
+      $NIX_BUILD --option experimental-features 'flakes' --impure --expr 'builtins.getFlake (builtins.toPath ./.).lib.testAgainst.x86_64-linux (builtins.getFlake (builtins.toPath ./.)).defaultPackage.x86_64-linux'
```
needs extra parens?
```suggestion
      $NIX_BUILD --option experimental-features 'flakes' --impure --expr '(builtins.getFlake (builtins.toPath ./.)).lib.testAgainst.x86_64-linux (builtins.getFlake (builtins.toPath ./.)).defaultPackage.x86_64-linux'
```

thufschmitt committed:
Fix the indentation of the yaml file

thufschmitt committed:
Add a missing pair of parens

thufschmitt committed:
Actually test against master

And not against ourself

thufschmitt committed:
Test using the current system, not an hardcoded one

thufschmitt committed:
Give the gh action access to a github token

Otherwise fetching the master branch will most likely fail because of
rate-limiting

thufschmitt committed:
Fetch Nix master from the cachix cache

Rather than rebuilding it each time

thufschmitt committed:
Run the tests as a separate flake

Avoids needing the `--impure` flag, and hopefully makes everything more
nicely cached

thufschmitt committed:
Use a github token if possible

To prevent some API limiting

thufschmitt committed:
Try to replicate as closely as possible the CI setup

So that the hashes match

regnat commented:
I’ve rewritten the test to make sure that we use exactly the same derivation that has been built previously (both for the current Nix daemon and master), so that we can just fetch it from the cache rather than re-building everything again. This brings the extra CI time to a bit less than 5mins (so not trivial, but acceptable)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-18/15300/1

Ericson2314 commented:
Oh this is tricky, but the goals is indeed a good one.

tomberek commented:
stale and OBE, reopen if needed
